core/trino-main/src/test/java/io/trino/operator/BenchmarkGroupByHashOnSimulatedData.java

```diff
@@ -0,0 +1,626 @@
+        private int[] channels;
+
+        public BenchmarkContext()
+        {
+            this.positions = DEFAULT_POSITIONS;
+        }
+
+        public BenchmarkContext(WorkType workType, AggregationDefinition query, double nullChance, int positions)
+        {
+            this.workType = requireNonNull(workType, "workType is null");
+            this.query = requireNonNull(query, "query is null");
+            this.positions = positions;
+            this.nullChance = nullChance;
+        }
+
+        @Setup
+        public void setup()
+        {
+            types = query.getChannels().stream()
+                    .map(channel -> channel.columnType.type)
+                    .collect(toImmutableList());
+            channels = IntStream.range(0, query.getChannels().size()).toArray();
+            pages = createPages(query);
+        }
+
+        private List<Page> createPages(AggregationDefinition definition)
+        {
+            List<Page> result = new ArrayList<>();
+            int channelCount = definition.getChannels().size();
+            int pageSize = definition.pageSize;
+            int pageCount = positions / pageSize;
+
+            Block[][] blocks = new Block[channelCount][];
+            for (int i = 0; i < definition.getChannels().size(); i++) {
+                ChannelDefinition channel = definition.getChannels().get(i);
+                blocks[i] = channel.createBlocks(pageCount, pageSize, i, nullChance);
+            }
+
+            for (int i = 0; i < pageCount; i++) {
+                int pageIndex = i;
+                Block[] pageBlocks = IntStream.range(0, channelCount)
+                        .mapToObj(channel -> blocks[channel][pageIndex])
+                        .toArray(Block[]::new);
+                result.add(new Page(pageBlocks));
+            }
+
+            return result;
+        }
+
+        public List<Page> getPages()
+        {
+            return pages;
+        }
+
+        public List<Type> getTypes()
+        {
+            return types;
+        }
+
+        public int[] getChannels()
+        {
+            return channels;
+        }
+
+        public WorkType getWorkType()
+        {
+            return workType;
+        }
+    }
+
+    public enum WorkType
+    {
+        ADD,
+        GET_GROUPS,
+    }
+
+    public enum AggregationDefinition
+    {
+        BIGINT_2_GROUPS(new ChannelDefinition(ColumnType.BIGINT, 2)),
+        BIGINT_10_GROUPS(new ChannelDefinition(ColumnType.BIGINT, 10)),
+        BIGINT_1K_GROUPS(new ChannelDefinition(ColumnType.BIGINT, 1000)),
+        BIGINT_10K_GROUPS(new ChannelDefinition(ColumnType.BIGINT, 10_000)),
+        BIGINT_100K_GROUPS(new ChannelDefinition(ColumnType.BIGINT, 100_000)),
+        BIGINT_1M_GROUPS(new ChannelDefinition(ColumnType.BIGINT, 1_000_000)),
+        BIGINT_10M_GROUPS(new ChannelDefinition(ColumnType.BIGINT, 10_000_000)),
+        BIGINT_2_GROUPS_1_SMALL_DICTIONARY(new ChannelDefinition(ColumnType.BIGINT, 2, 1, 50)),
```

Comment:
nit: add textual description to these enums (can be separate PR). It's hard to understand what they benchmark after some time

Reply:
https://github.com/trinodb/trino/pull/12170